### PR TITLE
Raw query editor with autocomplete

### PR DESCRIPTION
**Pull request by dprokop**
This brings a POC of a more generic SQLEditor component that would save us time implementing autocomplete for SQL-ish data sources over and over again.

The base of this PR has been set by @sunker with his autocomplete work done for the CloudWatch Datasource (ref grafana/cloudwatch-private#32)

This PR takes the core ideas implemented by Eric and generalizes those the same autocomplete mechanism can be used for multiple SQL data sources as well as extended to dialect-specific needs.

This is how it currently works with BigQuery data source, checkout this PR to see it live. Some syntax highlighting is broken due to the outdated Monaco version that we use:

https://user-images.githubusercontent.com/2376619/145589151-a8a27de7-9cd4-4c8d-b4b0-474fa31076d2.mov

🟢   This is how BigQuery completion provider implementation looks like: https://github.com/grafana/google-bigquery-datasource/blob/7353e7fa20a32aabb361a30507e9f39b2d2abe93/src/components/query-editor-raw/bigqueryCompletionProvider.ts It shows the most possible customization of the completion (custom suggestion kind, custom statement position resolver, tables & columns completion)

### Core concepts
- `SuggestionKind` - a descriptive string representing a type of a suggestion, i.e. `SelectKeyword`, `Tables`, `LogicalOperators` etc.
- `LinkedToken` - linked list element representing each individual token with a query. Allows traversing the query back and forth. Used by `StatementPositionResolver`(see below)
- `StatementPosition` - a desctiptive string representing cursor/token position within the query. Each statement position is defined together with `StatementPositionResolver` that, given some position context, returns a boolean value indicating whether or not we are in a given `StatementPosition` position. 
    ```ts
    export type StatementPositionResolver = (
        currentToken: LinkedToken | null,
        previousKeyword: LinkedToken | null,
        previousNonWhiteSpace: LinkedToken | null,
        previousIsSlash: Boolean // To be removed as it's CloudWatch specific
    ) => Boolean;
    ```
- `SuggestionKind` and `StatementPosition` are glued together via suggestions kind registry (language specific!). This registry contains items of `SuggestionKindRegistyItem` type of the following interface:
    ```ts
    export interface SuggestionKindRegistyItem extends RegistryItem {
        id: StatementPosition;
        kind: SuggestionKind[];
    }
    ```
    This item defines what kinds of suggestions should be provided in a given statement position
- Registries. There are couple of different registries used that drive the autocomplete mechanism. See the diagram[1] below to see how these work together:
  - **Language specific**: functions registry, operators registry, suggestion kinds registries and statement position resolvers registires. Those registires contain SQL defaults as well as allow extension per language type. 
  - **Instance specific**: Registry of `SuggestionsRegistyItem` items that glue particular `SuggestionKind` with an async function that provides completion items for it. 
    ```ts
    export interface SuggestionsRegistyItem extends RegistryItem {
        id: SuggestionKind;
        suggestions: (position: PositionContext, m: typeof monacoTypes) => Promise<CustomSuggestion[]>;
    }
    ```
    Think about instance-specific registry as having i.e. mixed data source with multiple query editors for the same type of data source and you wish to provide only table suggestions that are valid for particular query row.

### SQLEditor component

Goals
- [x] Allow providing suggestions for standard-ish SQL syntax (THIS PR)
- [x] Allow providing custom SQL dialects and suggestions for them (TODO - CloudWatch implementation sets a good base for how to provide custom dialect definition)

`SQLEditor` component builds on top of `CodeEditor` component, but we may want to base it on `ReactMonacoEditor` component instead to be less prone to `CodeEditor` API changes and have full controll over the Monaco API. For now the `CodeEditor` is good enough for a simplification.

`SQLEditor` API:

```ts
interface SQLEditorProps {
  query: string;
  onChange: (q: string) => void;
  language?: LanguageDefinition;
}
```

The important part is the `LanguageDefinition` interface which provides way to customize the completion both on a language and instance level:

```ts
interface LanguageDefinition extends monacoTypes.languages.ILanguageExtensionPoint {
  // TODO: Will allow providing a custom language definition.
  loadLanguage?: (module: any) => Promise<void>;
  // Provides API for customizing the autocomplete
  completionProvider?: (m: Monaco) => SQLCompletionItemProvider;
}
```

The `completionProvider` function is the core of the autocomplete customization. `SQLEditor` comes with standard SQL completion items, but this function allows:
- providing dynamic suggestions: tables, columns
- providing custom `StatementPositionResolvers` that are specific for a given dialect or not implemented yet for standard SQL
- providing custom `SuggestionKind` and resolvers for this kind of suggestions.

```ts
export interface SQLCompletionItemProvider
  extends Omit<monacoTypes.languages.CompletionItemProvider, 'provideCompletionItems'> {
  /**
   * Allows dialect specific functions to be added to the completion list.
   * @alpha
   */
  supportedFunctions?: () => Array<{
    id: string;
    name: string;
  }>;

  /**
   * Allows dialect specific operators to be added to the completion list.
   * @alpha
   */
  supportedOperators?: () => Array<{
    id: string;
    operator: string;
    type: OperatorType;
  }>;

  /**
   * Allows custom suggestion kinds to be defined and correlate them with <Custom>StatementPosition.
   * @alpha
   */
  customSuggestionKinds?: () => CustomSuggestionKind[];

  /**
   * Allows custom statement placement definition.
   * @alpha
   */
  customStatementPlacement?: () => CustomStatementPlacement[];

  /**
   * Allows providing a custom function for resolving db tables.
   * It's up to the consumer to decide whether the columns are resolved via API calls or preloaded in the query editor(i.e. full db schema is preloades loaded).
   * @alpha
   */
  tables?: {
    resolve: () => Promise<TableDefinition[]>;
    // Allows providing a custom function for calculating the table name from the query. If not specified a default implemnentation is used. I.e. BigQuery requires the table name to be fully qualified name: <project>.<dataset>.<table>
    parseName?: (t: LinkedToken) => string;
  };
  /**
   * Allows providing a custom function for resolving table.
   * It's up to the consumer to decide whether the columns are resolved via API calls or preloaded in the query editor(i.e. full db schema is preloades loaded).
   * @alpha
   */
  columns?: {
    resolve: (table: string) => Promise<ColumnDefinition[]>;
  };
}
```

## [1] Suggestions resolution flow and how API integrates with registries described above:

Untitled Diagram drawio (1)

### TODOs (not in prio order)
- [ ] ⚠️ Migrate tests from the CloudWatch implementation and adjust those against the generalized implementation
- [x] ⚠️  Move SQLEditor to grafana/experimental package
- [ ] Add support for macros autocomplete (default SQL macros + extension point for data source specif ones)
- [x] BUG: @grafana/core - update Monaco editor package to the latest one as some of the words are treated as SQL keywords even though they are not valid keywords (ref https://github.com/microsoft/monaco-editor/issues/2706) https://github.com/grafana/grafana/pull/43022
- [x] Custom `StatementPosition` resolvers
- [x] Provide API for resolving tables and columns
- [x] API for customizing available functions and operators
- [x] Nested `SELECT` - correctly resolve table name for columns retrieval
- [ ] Improve `WHERE` statement suggestions
- [ ] Improve `GROUP BY` statement suggestions
- [ ] Improve operators autocomplete
- [x] Think about a way to handle multiple query editors with the same language, yet different autocompletion (query specific) Proposal: 4386003cf7ce970d8280f2aee5846cccda68d922 🏃 
- [x] Support custom dialects definition 
- [ ] BUG: Typing first character into editor does not trigger suggestions (introduced with 4386003cf7ce970d8280f2aee5846cccda68d922)

### CODE REVIEW

**Review by sunker — COMMENTED** [review]
Fantastic work on this! In general, I think the abstractions look great. I really like the api for the SQLCodeEditor. 

A few thoughts from my side (not necessarily for this PR):
* I think we need to either minimize the standard options for all our registries **or** allow consumers to opt out on standard options.
* Consumers should be allowed to specify its own, custom monarch configuration so that it can be aligned with the custom auto complete experience they are providing. Currently, they can end up with syntax highlighting that don't match with the provided suggestions. 
* Loading language definition from the Monaco language repo adds little value, so I think we should keep a local copy in this lib
* The autocompletion should be able to handle schemas. e.g `select * from dbo.myTable`
* Missing autocomplete handling for a few standard things such as `join`, `having`, subqueries and wildcard (*) 
* One thing that I wasn't happy with in the CW autocomplete was testing. Ideally, we'd be able to use Monaco in our tests so that we can test the whole flow, i.e registering a language and then test autocompletion for a given query. However, I was not able to do that because currently Monaco is not built by webpack so you can't import it in your tests. I did not spend too much time on this, but I believe there should be ways to import the ESM directly into jest. If so, we should be able to write meaningful tests for this. 

Would be great if this draft could be moved to grafana-experimental sometime soon so that we could start using this in athena or redshift. By doing so, we'll be able to provide better feedback. 

>SQLEditor component builds on top of CodeEditor component, but we may want to base it on ReactMonacoEditor component instead to be less prone to CodeEditor API changes and have full controll over the Monaco API. For now the CodeEditor is good enough for a simplification.

@gabor a while back we talked about adding a new base component for react monaco that would only add basic styling, and would allow consumers to override any monaco props. Do you know if this work completed?

File: src/components/query-editor-raw/bigqueryCompletionProvider.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+              tablePath = processedToken.previous.value + tablePath;
+              processedToken = processedToken.previous;
+            }
+
+            const t = await getTables.current(tablePath);
+
+            return t.map((table) => ({
+              label: table.name,
+              insertText: table.completion ?? table.name,
+              kind: 1,
```
[the remark this answers is not in the document]
Comment (dprokop) [reply src/components/query-editor-raw/bigqueryCompletionProvider.ts:106]: It's not a SuggestionKind in this case, it's CompletionItemkind from monaco. But tru- this needs to be exposed :)

File: src/components/sql-editor/SQLEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    />
+  );
+};
+
+export const registerLanguageAndSuggestions = async (monaco: Monaco, l: LanguageDefinition, lid: string) => {
+  if (!l.loadLanguage) {
+    // Assume lanuage based on standard Monaco SQL definition
+    const defaultSQLLanguage = await getSQLLangConf(monaco);
+    monaco.languages.register({ id: lid });
+    monaco.languages.setMonarchTokensProvider(lid, { ...defaultSQLLanguage.language });
```
[the remark this answers is not in the document]
Comment (dprokop) [reply src/components/sql-editor/SQLEditor.tsx:82]: Yes - that's on the todos. This is just weryfying if stuff works in general :)

File: src/components/sql-editor/standardSql/suggestionsKindRegistry.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    },
+    {
+      id: StatementPosition.WhereComparisonOperator,
+      name: StatementPosition.WhereComparisonOperator,
+      kind: [SuggestionKind.ComparisonOperators],
+    },
+    {
+      id: StatementPosition.WhereValue,
+      name: StatementPosition.WhereValue,
+      kind: [SuggestionKind.LabelValues],
```
[the remark this answers is not in the document]
Comment (dprokop) [reply src/components/sql-editor/standardSql/suggestionsKindRegistry.ts:82]: Indeed - this kind of suggestion is not resolved anyways here and this is just a leftover.

**Comment by gabor** [comment]
@sunker  wrote: 
> @gabor a while back we talked about adding a new base component for react monaco that would only add basic styling, and would allow consumers to override any monaco props. Do you know if this work completed?

yes, it is done, it is the  `@grafana/ui/ReactMonacoEditor` widget. Prometheus query field uses it. it only sets up the theme based on grafana-theme, and nothing else. otherwise it's API is the same as `@monaco-editor/react`.

( i'd love to move it to `@grafana/experimental` eventually )

File: src/components/sql-editor/SQLEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+interface LanguageRegistries {
+  functions: Registry<FunctionsRegistryItem>;
+  operators: Registry<OperatorsRegistryItem>;
+  suggestionKinds: Registry<SuggestionKindRegistyItem>;
+  positionResolvers: Registry<StatementPositionResolversRegistryItem>;
+}
+
+const LANGUAGES_CACHE = new Map<string, LanguageRegistries>();
+const INSTANCE_CACHE = new Map<string, Registry<SuggestionsRegistyItem>>();
```
Comment (gabor) [inline src/components/sql-editor/SQLEditor.tsx:53]: for these two caches (`LANGUAGES_CACHE` and `INSTANCE_CACHE`).. do you plan to remove items from them at some point in some way, or they will just grow "forever"?

**Review by gabor — COMMENTED** [review]
i can only comment on the monaco-integration part, cannot comment on the how-to-sql-autocomplete, sorry.

one problem i faced with the prometheus-query-field too, was that you need to autocomplete based on instance-specific data.

i see you solved it by considering every instance of the editor-widget as a new language. this should work fine, the question is if it can cause problems in long-running systems if there will be hundreds of monaco-languages registered eventually (i would recommend unregistering unused languages at some point, but there does not seem to be an api to unregister languages in monaco 😿 ).

in prometheus i handled this in a different way, maybe it would help here, maybe not, i don't know, i'll just describe it here:
in https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/components/monaco-query-field/MonacoQueryField.tsx#L162-L166

for prometheus we only register one monaco-language, and basically we setup every autocomplete-callback on that same language.
but the autocomplete-function we register is just a proxy to the REAL autocomplete-function, and it does this:
- was i called from the editor to which i "really belong"? if not, return empty-list as suggestions
- otherwise call the "real" autocomplete function

it is achieved by storing the `onMount` `editor` instance, so we can find out if it is the correct editor or not, in the autocomplete callback.

this might be a bad choice for the sql-editor, i don't know, i just wanted to describe an alternative approach.

File: src/components/sql-editor/SQLEditor.tsx (hunk 4 above)
Comment (dprokop) [reply src/components/sql-editor/SQLEditor.tsx]: Yes, I plan to clean those caches

**Comment by dprokop** [comment]
> One thing that I wasn't happy with in the CW autocomplete was testing. Ideally, we'd be able to use Monaco in our tests so that we can test the whole flow, i.e registering a language and then test autocompletion for a given query. However, I was not able to do that because currently Monaco is not built by webpack so you can't import it in your tests. I did not spend too much time on this, but I believe there should be ways to import the ESM directly into jest. If so, we should be able to write meaningful tests for this.

@sunker indeed - I was struggling to get monaco working in jest env as well. Would be just perfect if we could use the tokenizer in the test rather than providing the test queries with tokens, it's a PITA.

File: src/components/sql-editor/SQLEditor.tsx (hunk 4 above)
Comment (dprokop) [reply src/components/sql-editor/SQLEditor.tsx]: https://github.com/grafana/grafana-experimental/pull/9

File: src/components/sql-editor/SQLEditor.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (dprokop) [reply src/components/sql-editor/SQLEditor.tsx]: Fixed https://github.com/grafana/grafana-experimental/pull/10

**Review by zoltanbedi — APPROVED** [review]
I took this for a spin and it works great. We can merge this and work on the cleanup in separate prs if that makes sense to you.

**Comment by dprokop** [comment]
> I took this for a spin and it works great. We can merge this and work on the cleanup in separate prs if that makes sense to you.

That's exactly my plan @zoltanbedi , about to merge in a few :)

**Pull request merged**